xarray/core/coordinates.py

```diff
@@ -796,6 +797,126 @@ def _ipython_key_completions_(self):
         ]
 
 
+class DataTreeCoordinates(DatasetCoordinates):
+    """
+    Dictionary like container for coordinates of a DataTree node (variables + indexes).
+    """
+
+    # TODO: This collection can be passed directly to the :py:class:`~xarray.Dataset`
+    # and :py:class:`~xarray.DataArray` constructors via their `coords` argument.
+    # This will add both the coordinates variables and their index.
+
+    # TODO: This only needs to be a separate class from `DatasetCoordinates` because DataTree nodes store their variables differently
+    # internally than how Datasets do, see https://github.com/pydata/xarray/issues/9203.
+
+    # TODO should inherited coordinates be here? It would be very hard to allow updating them...
+    # But actually maybe the ChainMap approach would make this work okay??
+
+    _data: DataTree
+
+    __slots__ = ("_data",)
+
+    def __init__(self, datatree: DataTree):
+        self._data = datatree
+
+    @property
+    def _names(self) -> set[Hashable]:
+        return set(self._data._coord_variables)
+
+    @property
+    def dims(self) -> Frozen[Hashable, int]:
+        # TODO is there a potential bug here? What happens if a dim is only present on data variables?
+        return self._data.dims
```

Comment:
I think this should be fine. Dimensions on `DatasetCoordinates` also include dimensions that are only present on data variables.

Reply:
Really? This (on `main`) looks wrong to me:

```python
In [1]: import xarray as xr

In [2]: ds = xr.Dataset({'a': ('x', [0, 1])})

In [3]: ds
Out[3]: 
<xarray.Dataset> Size: 16B
Dimensions:  (x: 2)
Dimensions without coordinates: x
Data variables:
    a        (x) int64 16B 0 1

In [4]: ds.coords
Out[4]: 
Coordinates:
    *empty*

In [5]: ds.coords.dims
Out[5]: FrozenMappingWarningOnValuesAccess({'x': 2})
```
I mean the fact no-one has raised this before means it probably isn't of much consequence, but it does seem incorrect / misleading.

Reply:
Raised https://github.com/pydata/xarray/issues/9466 to track this